### PR TITLE
Ignore ~/.cabal if $XDG_CONFIG_HOME/cabal/config exists.

### PR DESCRIPTION
**Pull request by athas**
This is a draft implementation of #8577.

There is one flaw in this implementation: the warning is printed many times, and with non-configurable verbosity.  Changing that would require one of:

1) Significant refactoring of how `getDefaultDir` is called, which has non-local implications.
2) Doing the warning somewhere else (e.g. in Main.hs).

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by athas** [comment]
It now warns only when `loadConfig` is called, but for some commands (e.g. `cabal install`) that is still twice.

File: cabal-install/src/Distribution/Client/Config.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/Config.hs:608]: The parenthesis are confusing. How about:

> If @CABAL\_DIR@ is set return @Just@ its value. Otherwise, if @~/.cabal@ exists and
> @$XDG_CONFIG_HOME/cabal/config@ does not exist, return @Just "~/.cabal"@. 
> Otherwise, return @Nothing@.

File: cabal-install/src/Distribution/Client/Config.hs
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/Config.hs:606]: How about:

> … ignoring the former. It is advisable to remove one of them. In that case, we will use the remaining one by default (unless '$CABAL_DIR' is explicitly specified).

**Review by ulysses4ever — APPROVED** [review]
It’s my understanding that there’s no way we can test that it does the right thing, is there?

**Comment by ulysses4ever** [comment]
I’m not worried about repetitive warnings: it won’t happen too often, and really the user should make sure that there’s only one thing exists.

**Comment by athas** [comment]
> It’s my understanding that there’s no way we can test that it does the right thing, is there?

In principle it is possible to do so in an ad-hoc manner (after all, I tested it manually), but I don't know whether cabal's test suite is set up to make it easy.

**Comment by Mikolaj** [comment]
What's the status here? Let me add a needs-review label in case it's ready for review...

**Comment by athas** [comment]
I don't remember anything left to do here.  There are no tests, though.

**Review by gbaz — APPROVED** [review]

**Comment by Kleidukos** [comment]
@athas If you could write down some instructions for QA testers so that they can manually check your patch, that would be grand. :)

**Comment by athas** [comment]
Sure.  There are the following situations to test, based on which files and directories exist.  Their contents do not matter.

1. Neither `$HOME/.cabal` nor `$HOME/.config/cabal/config` exist.  In this case `cabal help` should print `$HOME/.config/cabal/config` on the second-to-last line.
2. `$HOME/.cabal` exists as a directory and `$HOME/.config/cabal/config` does not exist.  `cabal help` should print `$HOME/.config/cabal/config` on the second-to-last line (or last line if `$HOME/.cabal/config` actually exists).
3. `$HOME/.cabal` exists as a directory but `$HOME/.config/cabal/config` also exists.  `cabal help` should print `$HOME/.config/cabal/config` on the last line.
4. `$HOME/.cabal` does not exist but `$HOME/.config/cabal/config` exists.  `cabal help` should print `$HOME/.config/cabal/config` on the last line.

**Comment by Mikolaj** [comment]
@athas: it's normally PR author's role, but let me set the label in the interest of merging this, before CI breaks or something. :) This will also unblock QA, which happens after the merge.

**Comment by Mikolaj** [comment]
@mergify backport 3.10

**Comment by Mikolaj** [comment]
@mergify unqueue

**Comment by Mikolaj** [comment]
@mergify queue

**Comment by ulysses4ever** [comment]
@mergify refresh

**Comment by ulysses4ever** [comment]
There's a funny [Mergify failure here](https://github.com/haskell/cabal/pull/8877/checks?check_run_id=13724984175) : 

```
The merge queue pull request can't be updated
Details:

    Base branch update has failed

    refusing to allow a GitHub App to create or update workflow .github/workflows/quick-jobs.yml without workflows permission
    err-code: CDDDA
```

I don't understand why Mergify (apparently) tries to modify the workflow file...

**Comment by ulysses4ever** [comment]
Googling gives this: https://github.com/Mergifyio/mergify/issues/5055 supposedly resolved by this comment:

> This is a GitHub App permission problem. If you log into your Mergify dashboard you should see a message requesting you to update the permission for the app.

I don't see anything like that in the dashboard but on one of the tabs it says that I should be an GH Org admin to fiddle with "Application keys", which I'm not sure whether it's related or not. https://dashboard.mergify.com/github/haskell/application-keys

**Comment by athas** [comment]
Is something more expected of me?  It's not that I would mind writing automated tests, but someone gave me the impression that it would not be easy to fit into cabal's current test infrastructure.  I'm fine with adding some more furniture, but I don't currently have the time to rework the plumbing.

**Comment by Mikolaj** [comment]
@mergify rebase

**Comment by ulysses4ever** [comment]
@athas I think your job is done here, thanks a lot! Let's keep in touch: XDG is a rather fundamental change, and I expect more questions will need to be discussed in the future.

**Comment by jgotoh** [comment]
Manual QA succeeded for the provided steps, if anyone is interested here is the console output:
https://gist.github.com/jgotoh/1626aba91f742b74629c689a9af827d6

**Comment by ulysses4ever** [comment]
Great! 🎉

**Comment by Mikolaj** [comment]
Thanks, this works for me in the semi-wild with cabal 3.10.2.0-alpha1. I had no idea I have cabal configs in both locations and it detected it. Here's my battle report from the Matrix channel:

------

FYI, I'm dog-fooding the centos7 flavour of cabal 3.10.2.0-alpha1 from https://gitlab.haskell.org/haskell/cabal/-/jobs/1651797/artifacts/browse/out/
so far, so good, it builds stuff fine (though there is a slight inconsistency in the message below, as it mentions a directory first and a (config) file in an analogous directory second)

```
123456789~/r/horde-ad$ cabal build horde-ad-simplified --enable-optimization 
Warning: Both /home/mikolaj/.cabal and /home/mikolaj/.config/cabal/config
exist - ignoring the former.
It is advisable to remove one of them. In that case, we will use the remaining
one by default (unless '$CABAL_DIR' is explicitly set).
Warning: Both /home/mikolaj/.cabal and /home/mikolaj/.config/cabal/config
exist - ignoring the former.
It is advisable to remove one of them. In that case, we will use the remaining
one by default (unless '$CABAL_DIR' is explicitly set).
```

(and the message is repeated twice, but that's probably unavoidable; IDK)
in any case, it detected my duplicated cabal config and warned me, so it prevents real user problems

**Comment by Kleidukos** [comment]
We have the option of mentioning `$CABAL_CONFIG` instead.

We can also link to the manual: https://cabal.readthedocs.io/en/stable/config.html#configuration-file-discovery

**Comment by Mikolaj** [comment]
If I'm wrong the message duplication is easy to fix, then a patch would be helpful. Then we could also make the mentions of directory `/home/mikolaj/.cabal` vs file `/home/mikolaj/.config/cabal/config` consistent and mention $CABAL_CONFIG instead or in addition to $CABAL_DIR (I'm happily ignorant about what they do).

I think there's absolutely no rush. The problems are minor and the changes will be backward compatible, so they don't need to be included this release. But I may be misunderstanding any of the above.

**Pull request merged**
